File: app_manager/appmanager/src/main/java/app/editors/manager/ui/activities/main/OperationActivity.java
package app.editors.manager.ui.activities.main;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;

import com.google.android.material.appbar.AppBarLayout;

import javax.inject.Inject;

import app.editors.manager.R;
import app.editors.manager.app.Api;
import app.editors.manager.app.App;
import app.editors.manager.app.WebDavApi;
import app.editors.manager.managers.tools.PreferenceTool;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.mvp.models.explorer.Explorer;
import app.editors.manager.mvp.models.states.OperationsState;
import app.editors.manager.ui.activities.base.BaseAppActivity;
import app.editors.manager.ui.fragments.operations.DocsCloudOperationFragment;
import app.editors.manager.ui.fragments.operations.DocsOperationSectionFragment;
import app.editors.manager.ui.fragments.operations.DocsWebDavOperationFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class OperationActivity extends BaseAppActivity {

    public static final String TAG = OperationActivity.class.getSimpleName();
    public static final String TAG_OPERATION_TYPE = "TAG_OPERATION_OPERATION_TYPE";
    public static final String TAG_OPERATION_EXPLORER = "TAG_OPERATION_EXPLORER";
    public static final String TAG_IS_WEB_DAV = "TAG_IS_WEB_DAV";

    public interface OnActionClickListener {
        void onActionClick();
    }

    private Unbinder mUnbinder;
    @BindView(R.id.app_layout)
    protected CoordinatorLayout mAppLayout;
    @BindView(R.id.app_bar_layout)
    protected AppBarLayout mAppBarLayout;
    @BindView(R.id.app_bar_toolbar)
    protected Toolbar mAppBarToolbar;
    @BindView(R.id.operation_cancel_button)
    protected AppCompatButton mOperationCancelButton;
    @BindView(R.id.operation_action_button)
    protected AppCompatButton mOperationActionButton;

    @Inject
    protected PreferenceTool mPreferenceTool;

    private OperationsState.OperationType mOperationType;
    private OnActionClickListener mOnActionClickListener;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.getApp().getAppComponent().inject(this);
        setContentView(R.layout.activity_operation);
        mUnbinder = ButterKnife.bind(this);
        init(savedInstanceState);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mUnbinder.unbind();
    }

    @OnClick({R.id.operation_cancel_button, R.id.operation_action_button})
    protected void onButtonClick(final View view) {
        switch (view.getId()) {
            case R.id.operation_cancel_button:
                setResult(Activity.RESULT_CANCELED);
                finish();
                break;
            case R.id.operation_action_button:
                if (mOnActionClickListener != null) {
                    mOnActionClickListener.onActionClick();
                }
                break;
        }
    }

    private void init(final Bundle savedInstanceState) {
        setFinishOnTouchOutside(true);
        setSupportActionBar(mAppBarToolbar);
        mOperationType = (OperationsState.OperationType) getIntent().getSerializableExtra(OperationActivity.TAG_OPERATION_TYPE);
        initButton(mOperationType);

        if (savedInstanceState == null) {
            if (mPreferenceTool.isPersonalPortal()) {
                showFragment(DocsCloudOperationFragment.newInstance(Api.SectionType.CLOUD_USER), null);
            } else if (App.getApp().getAppComponent().getAccountsSql().getAccountOnline().isWebDav()) {
                AccountsSqlData account = App.getApp().getAppComponent().getAccountsSql().getAccountOnline();
                showFragment(DocsWebDavOperationFragment.newInstance(WebDavApi.Providers.valueOf(account.getWebDavProvider())), null);
            } else {
                showFragment(DocsOperationSectionFragment.newInstance(), null);
            }
        }
    }

    private void initButton(OperationsState.OperationType actionOperationType) {
        switch (actionOperationType) {
            case COPY:
                mOperationActionButton.setText(R.string.operation_panel_copy_button);
                break;
            case MOVE:
                mOperationActionButton.setText(R.string.operation_panel_move_button);
                break;
        }
    }

    public void setEnabledActionButton(final boolean isEnabled) {
        mOperationActionButton.setEnabled(isEnabled);
    }

    public void setOnActionClickListener(OnActionClickListener onActionClickListener) {
        mOnActionClickListener = onActionClickListener;
    }

    public static void showCopy(final Fragment fragment, @NonNull Explorer explorer) {
        final Intent intent = new Intent(fragment.getContext(), OperationActivity.class);
        intent.putExtra(OperationActivity.TAG_OPERATION_TYPE, OperationsState.OperationType.COPY);
        intent.putExtra(OperationActivity.TAG_OPERATION_EXPLORER, explorer);
        fragment.startActivityForResult(intent, REQUEST_ACTIVITY_OPERATION);
    }

    public static void showMove(final Fragment fragment, @NonNull Explorer explorer) {
        final Intent intent = new Intent(fragment.getContext(), OperationActivity.class);
        intent.putExtra(OperationActivity.TAG_OPERATION_TYPE, OperationsState.OperationType.MOVE);
        intent.putExtra(OperationActivity.TAG_OPERATION_EXPLORER, explorer);
        fragment.startActivityForResult(intent, REQUEST_ACTIVITY_OPERATION);
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/activities/main/OperationActivity.kt
package app.editors.manager.ui.activities.main

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.lifecycle.lifecycleScope
import app.documents.core.database.datasource.CloudDataSource
import app.documents.core.model.cloud.WebdavProvider
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.explorer.Explorer
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.app.accountOnline
import app.editors.manager.databinding.ActivityOperationBinding
import app.editors.manager.mvp.models.states.OperationsState.OperationType
import app.editors.manager.ui.activities.base.BaseAppActivity
import app.editors.manager.ui.fragments.operations.DocsCloudOperationFragment
import app.editors.manager.ui.fragments.operations.DocsDropboxOperationFragment
import app.editors.manager.ui.fragments.operations.DocsGoogleDriveOperationFragment
import app.editors.manager.ui.fragments.operations.DocsOneDriveOperationFragment
import app.editors.manager.ui.fragments.operations.DocsOperationSectionFragment
import app.editors.manager.ui.fragments.operations.DocsWebDavOperationFragment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import lib.toolkit.base.managers.utils.getSerializable
import javax.inject.Inject

class OperationActivity : BaseAppActivity(){

    interface OnActionClickListener {
        fun onActionClick()
    }

    companion object {
        val TAG: String = OperationActivity::class.java.simpleName

        const val TAG_OPERATION_TYPE = "TAG_OPERATION_OPERATION_TYPE"
        const val TAG_OPERATION_EXPLORER = "TAG_OPERATION_EXPLORER"
        const val TAG_IS_WEB_DAV = "TAG_IS_WEB_DAV"

        fun getIntent(context: Context, operation: OperationType, explorer: Explorer) = Intent(context, OperationActivity::class.java).apply {
            putExtra(TAG_OPERATION_TYPE, operation)
            putExtra(TAG_OPERATION_EXPLORER, explorer)
        }
    }

    @Inject
    lateinit var cloudDataSource: CloudDataSource

    private var operationType: OperationType? = null
    private var actionClickListener: OnActionClickListener? = null
    private var viewBinding: ActivityOperationBinding? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.getApp().appComponent.inject(this)
        viewBinding = ActivityOperationBinding.inflate(layoutInflater)
        setContentView(viewBinding?.root)
        init(savedInstanceState)
        initListeners()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        viewBinding = null
    }

    private fun init(savedInstanceState: Bundle?) {
        setFinishOnTouchOutside(true)
        setSupportActionBar(viewBinding?.appBarToolbar)
        operationType = intent.getSerializable(TAG_OPERATION_TYPE, OperationType::class.java)
        initButton(operationType)
        if (savedInstanceState == null) {
            initState()
        }
    }

    private fun initState() {
       lifecycleScope.launch {
            accountOnline?.let { account ->
                withContext(Dispatchers.Main) {
                    if (account.isWebDav) {
                        showFragment(
                            DocsWebDavOperationFragment.newInstance(
                                WebdavProvider.valueOf(account.portal.provider)
                            ), null
                        )
                    } else if(account.isOneDrive) {
                        showFragment(DocsOneDriveOperationFragment.newInstance(), null)
                    } else if(account.isDropbox) {
                        showFragment(DocsDropboxOperationFragment.newInstance(), null)
                    } else if(account.isGoogleDrive) {
                        showFragment(DocsGoogleDriveOperationFragment.newInstance(), null)
                    } else {
                        if (account.isPersonal()) {
                            showFragment(DocsCloudOperationFragment.newInstance(ApiContract.SectionType.CLOUD_USER), null)
                        } else {
                            showFragment(DocsOperationSectionFragment.newInstance(Json.encodeToString(account)), null)
                        }
                    }
                }
            }
        }
    }

    private fun initListeners() {
        viewBinding?.operationPanel?.operationActionButton?.setOnClickListener {
            actionClickListener?.onActionClick()
        }
        viewBinding?.operationPanel?.operationCancelButton?.setOnClickListener {
            setResult(RESULT_CANCELED)
            finish()
        }
    }

    private fun initButton(actionOperationType: OperationType?) {
        when (actionOperationType) {
            OperationType.COPY -> viewBinding?.operationPanel?.operationActionButton?.setText(R.string.operation_panel_copy_button)
            OperationType.MOVE -> viewBinding?.operationPanel?.operationActionButton?.setText(R.string.operation_panel_move_button)
            OperationType.RESTORE -> viewBinding?.operationPanel?.operationActionButton?.setText(R.string.operation_panel_restore_button)
            else -> {
            }
        }
    }

    fun setEnabledActionButton(isEnabled: Boolean) {
        viewBinding?.operationPanel?.operationActionButton?.isEnabled = isEnabled
    }

    fun setOnActionClickListener(onActionClickListener: OnActionClickListener?) {
        actionClickListener = onActionClickListener
    }

}